modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java

```diff
@@ -265,14 +265,29 @@ private RelNode convertMerge(SqlMerge call) {
             // provided, in which case, the expression is the default value for
             // the column; or if the expressions directly map to the source
             // table
-            level1InsertExprs =
-                    ((LogicalProject) insertRel.getInput(0)).getProjects();
-            if (insertRel.getInput(0).getInput(0) instanceof LogicalProject) {
-                level2InsertExprs =
-                        ((LogicalProject) insertRel.getInput(0).getInput(0))
-                                .getProjects();
+            RelNode input = insertRel.getInput(0);
+
+            if (input instanceof LogicalProject) {
+                level1InsertExprs = ((LogicalProject) input).getProjects();
+            } else {
+                // TODO https://issues.apache.org/jira/browse/IGNITE-22293
+                // convertInsert() may return LogicalTableModify without projection in the input.
+                // As a workaround for this case, we additionally build required column expressions.
+                RelDataType rowType = input.getRowType();
+                RexNode sourceRef = rexBuilder.makeRangeReference(rowType, 0, false);
+                ArrayList<String> targetColumnNames = new ArrayList<>(rowType.getFieldCount());
+                ArrayList<RexNode> columnExprs = new ArrayList<>(rowType.getFieldCount());
+
+                collectInsertTargets(insertCall, sourceRef, targetColumnNames, columnExprs);
```

Comment:
seems all this stuff can be changed for more lightweight:
```
                ArrayList<RexNode> columnExprs = new ArrayList<>(rowType.getFieldCount());
                int pos = 0;
                for (RelDataTypeField type : rowType.getFieldList()) {
                    columnExprs.add(rexBuilder.makeInputRef(type.getType(), pos++));
                }
```

Reply:
Fixed, thank you.